src/libraries/System.Text.Json/docs/ReferenceHandling_spec.md

```diff
@@ -0,0 +1,810 @@
+
+* Having more than one `$id` in the same object:
+  * **Newtonsoft.Json**: last one wins, in the example, the reference object evaluates to `null` (if `$ref` would be `"2"`, it would evaluate to itself).
+  * **S.T.Json**: Throw - Object already defines a reference identifier.
+```json
+{
+    "$id": "1",
+    "$id": "2",
+    "Name": "Angela",
+    "Manager": {
+        "$ref": "1"
+    }
+}
+```
+
+* `$id` is not the first property:
+  * **Newtonsoft.Json**: Object is not preserved and cannot be referenced, therefore any reference to it would evaluate as null.
+
+  * **S.T.Json**: Throw - Object $id is not the first property.
+  Note: In case we would want to switch, we can handle the `$id` not being the first property since we store the reference at the moment we spot the `$id` property, we throw to honor the rule of thumb.
+```json
+{
+    "Name": "Angela",
+    "$id": "1",
+    "Manager": {
+        "$ref": "1"
+    }
+}
+```
+
+* `$id` is duplicated (not necessarily nested):
+  * **Newtonsoft.Json**: Throws - Error reading object reference '1'- Inner Exception: ArgumentException: A different value already has the Id '1'.
+  * **S.T.Json**: Throws - Duplicated id found while preserving reference.
+```json
+[
+    {
+        "$id": "1",
+        "Name": "Angela"
+    },
+    {
+        "$id": "1",
+        "Name": "Bob"
+    }
+]
+```
+
+## Preserved arrays
+A regular array is `[ elem1, elem2 ]`.
+A preserved array is written in the next format `{ "$id": "1", "$values": [ elem1, elem2 ] }`
+
+* Preserved array does not contain any metadata:
+  * **Newtonsoft.Json**: Throws - Cannot deserialize the current JSON object into type 'System.Collections.Generic.List`1
+  * **S.T.Json**: Throw - Preserved array $values property was not present or its value is not an array.
+
+  ```json
+  {}
+  ```
+
+* Preserved array only contains $id:
+  * **Newtonsoft.Json**: Throws - Cannot deserialize the current JSON object into type 'System.Collections.Generic.List`1
+  * **S.T.Json**: Throw - Preserved array $values property was not present or its value is not an array.
+
+  ```json
+  {
+      "$id": "1"
+  }
+  ```
+
+* Preserved array only contains `$values`:
+  * **Newtonsoft.Json**: Does not throw and the payload evaluates to the array in the property.
+  * **S.T.Json**: Throw - Preserved arrays cannot lack an identifier.
+
+  ```json
+  {
+      "$values": []
+  }
+  ```
+
+* Preserved array $values property contains null
+  * **Newtonsoft.Json**: Throw - Unexpected token while deserializing object: EndObject. Path ''.
+  * **S.T.Json**: Throw - Preserved array $values property was not present or its value is not an array.
+
+  ```json
+  {
+      "$id": "1",
+      "$values": null
+  }
+  ```
+
+* Preserved array $values property contains value
+  * **Newtonsoft.Json**: Unexpected token while deserializing object: EndObject. Path ''.
+  * **S.T.Json**: Throw - The JSON value could not be converted to TArray. Path: $.$values
```

Comment:
Why is the error message here different than the `null` case above?

Reply:
Because this error is the same as if you were trying to set a primitive (which uses converter) into a collection. I didn't want to add a condition into the converter that would check for every primitive type. It sounds like too much overhead.

Reply:
```suggestion
  * **S.T.Json**: Throw - Preserved array $values property was not present or its value is not an array.
```